### PR TITLE
quincy:  mon, osd: rework the public_bind_addr support. Bring it to OSD

### PR DESCRIPTION
**Pull request by rzarzynski**
This is the quincy backport of PR https://github.com/ceph/ceph/pull/49990.

Backport ticket: https://tracker.ceph.com/issues/58781.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/common/pick_address.cc [hunk 1]
```diff
@@ -354,10 +355,11 @@ int pick_addresses(
   addrs->v.clear();
 
   unsigned addrt = (flags & (CEPH_PICK_ADDRESS_PUBLIC |
+			     CEPH_PICK_ADDRESS_PUBLIC_BIND |
 			     CEPH_PICK_ADDRESS_CLUSTER));
-  if (addrt == 0 ||
-      addrt == (CEPH_PICK_ADDRESS_PUBLIC |
-		CEPH_PICK_ADDRESS_CLUSTER)) {
+  if (std::popcount(addrt) != 1) {
```
Comment (ronen-fr) [inline src/common/pick_address.cc:360]: Might want to consider std::has_single_bit()?

**Comment by ronen-fr** [comment]
2'nd commit msg: "Monitors do" or "A... does"?

**Review by ronen-fr — APPROVED** [review]

File: src/common/pick_address.cc (hunk 1 above)
Comment (idryomov) [reply src/common/pick_address.cc:360]: `std::has_single_bit` is C++20 which is not enabled for quincy.

**Comment by idryomov** [comment]
> 2'nd commit msg: "Monitors do" or "A... does"? Also should be 'it' in the last paragraph, and 'is based' & 'still' in the second

The main PR isn't merged yet, it can be fixed there and carried over here.

File: src/common/pick_address.cc (hunk 1 above)
Comment (idryomov) [reply src/common/pick_address.cc:360]: Argh, `std::popcount()` is also C++20 and that's why this backport doesn't compile!  I thought it was C++17 for some reason...

Comment (ronen-fr) [reply src/common/pick_address.cc:360]: but so is popcount()...

Comment (idryomov) [reply src/common/pick_address.cc:360]: Yeah -- our comments raced ;)  Whichever Radek chooses to use in the main PR, the backport is going to need to resort to a builtin.

**Comment by rzarzynski** [comment]
Fixed the commit description and switched-out from `std::popcount` to pretty ugly `std::bitset<...>::count`. Thanks for the reviews!

**Review by idryomov — APPROVED** [review]

**Pull request merged**